### DISCUSSION TITLE
Rewrite the registry being used for docker dependencies

### DISCUSSION THREAD

**Discussion by knechtionscoding** [renovatebot/renovate#22795] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.61.0

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

In our current configuration we have an ECR repository for each region/account where our k8s clusters are pulled from. In order to avoid having to give Renovate access to each one of those, we'd like to have renovate (using helm-values) check a central repository where all images are pulled in first. 

I.e. we have a bunch of references to `repository-1.dkr.ecr.eu-central-1.amazonaws.com` and `repository-2.dkr.ecr.eu-west-1.amazonaws.com` and we would like to tell renovate to use a corp.registry.com to do the lookup for a new version. Resulting in a lookup of: `corp.registry.com/nginx:1.0.0`

### What you tried so far.

I've attempted to exclude the registry completely:

```yaml
image:
  name: test_repo/nginx
  tag: 1.0.0
```

With the package rule:

```json
          "packageRules": [
            {
              "matchDatasources": ["docker"],
              "matchDepPatterns": ".*test_repo.*",
              "registryUrls":["https://corp.registry.com"],
            }
```

This works, except that it gets rewritten to: `corp.registry.com/test_repo` which obviously isn't found.

I then tried:

```yaml
image:
  name: test.local/nginx
  tag: 1.0.0
```

with the package rule:

```json
          "packageRules": [
            {
              "matchDatasources": ["docker"],
              "matchPackagePatterns": ".*test.local.*",
              "registryUrls":["https://corp.registry.com"],
            }
```

This doesn't appear to work. Resulting instead in: 

```
{"name":"renovate","hostname":"renovate-28113915-wtfrj","pid":9,"level":20,"logContext":"Og9a3VhFK-i3mz8AWSPLr","repository":"cool-repo","dependency":"registry-1.dkr.ecr.eu-central-1.amazonaws.com/nginx:1.0.0","packageFile":"versions.yaml","msg":"Failed to look up docker package registry-1.dkr.ecr.eu-central-1.amazonaws.com/nginx:1.0.0","time":"2023-06-15T13:18:33.899Z","v":0}
```

### Relevant debug logs

<details><summary>Logs</summary>

```
{"name":"renovate","hostname":"renovate-28113915-wtfrj","pid":9,"level":20,"logContext":"Og9a3VhFK-i3mz8AWSPLr","repository":"cool-repo","dependency":"registry-1.dkr.ecr.eu-central-1.amazonaws.com/nginx:1.0.0","packageFile":"versions.yaml","msg":"Failed to look up docker package registry-1.dkr.ecr.eu-central-1.amazonaws.com/nginx:1.0.0","time":"2023-06-15T13:18:33.899Z","v":0}
```

</details>

**Comment by rarkins** [comment, marked as answer]
Start by checking your `packageFiles with updates` debug log message do see what the depName and packageName are for the one you're trying to rewrite.

IIRC `registryUrls` don't work when the image name is already fully qualified (with a host name). I don't recall if `registryAliases` can be used instead - it's a possibility

**Reply by viceice** [reply]
not possible
- #17963

https://github.com/renovatebot/renovate/blob/7dd8845bb013d8db2bada7d19e95d13435e0dec8/lib/config/validation.ts#L57-L64

**Reply by viceice** [reply]
- #22815

**Reply by knechtionscoding** [reply]
Thanks! I'll take a look at registryAlias. There's not a ton of documentation on the expected use. I see the discussion of $CI_REGISTRY being replaced.

For registry aliases to work In the yaml file where versions are declared, should it be listed as: `$CI_REGISTRY/nginx:1.0.0` or as `test.local/nginx:1.0.0` or just `nginx:1.0.0`? What's the syntax supposed to be like

**Reply by knechtionscoding** [reply]
And follow-up, I see that 35.122.0 was released containing loosening restrictions on the validation (thanks @viceice!) but I don't see that as a docker tag: https://hub.docker.com/r/renovate/renovate/tags I do see the github release. Do I just need to be patient for it to show up?

Edit: It is now available

**Reply by viceice** [reply]
it's just a string/ string mapping. so we do a simple `string.replace(key,value)`

**Reply by viceice** [reply]
so you can effectively replace any string inside the image name.

**Reply by knechtionscoding** [reply]
Apologies for the delay, was off.

Trying: 

```json
            {
              "groupName": "RegistryAlias",
              "matchPackagePatterns": "^fake-docker.example\/.*",
              "registryAliases": {
                "fake-docker.example": "corp.registry.com"
              }
            },
```

Which results in the logline:

```
{"name":"renovate","hostname":"renovate-28120965-fwf5w","pid":9,"level":20,"logContext":"S2uWIUKtCCR17oh_68RfJ","repository":"hknecht/cloud-services-infra","dependency":"fake-docker.example/nginx","packageFile":"versions.yaml","msg":"Failed to look up docker package fake-docker.example/nginx","time":"2023-06-20T10:47:33.162Z","v":0}
```

**Reply by viceice** [reply]
aliases need to be global in renovate JSON, can't be used inside package rules

**Reply by knechtionscoding** [reply]
Makes sense. Thanks!
